### ISSUE TITLE
The quality of generated image using NeuronStableDiffusionPipeline is very poor

### ISSUE THREAD

**Issue by newgrit1004** [huggingface/optimum-neuron#607]
### System Info

```shell
AMI :Deep Learning AMI Neuron PyTorch 1.13 (Ubuntu 20.04) 20240411
python : 3.8.10

[libraries]
optimum-neuron 0.0.22
torch 2.1.2 
torch-neuronx 2.1.2.2.1.0
torch-xla 2.1.2 
torchvision 0.16.2
transformers 4.36.2
transformers-neuronx 0.10.0.21
triton 2.1.0
diffusers 0.27.2
```

### Who can help?

@JingyaHuang 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction (minimal, reproducible, runnable)

I followed the part "Export via Python API" in [this link](https://huggingface.co/docs/optimum-neuron/tutorials/stable_diffusion#compile-stable-diffusion) to compile the neuron model

```python
from optimum.neuron import NeuronStableDiffusionPipeline
model_id = "runwayml/stable-diffusion-v1-5"
compiler_args = {"auto_cast": "matmul", "auto_cast_type": "bf16"}
input_shapes = {"batch_size": 1, "height": 512, "width": 512}
stable_diffusion = NeuronStableDiffusionPipeline.from_pretrained(model_id, export=True, **compiler_args, **input_shapes)
save_directory = "sd_neuron/"
stable_diffusion.save_pretrained(save_directory)
```

Then I want to compare the generated images between original stable diffusion pipeline and neuron stable diffusion pipeline.
So, I run the code below for the benchmark.
Since cuda is not available in inferentia instance, I run the original stable diffusion pipeline using cpu to set the same generator with same seed.

```python
import torch
import time
def example_prompts():
    prompts = [
        "a photo of an astronaut riding a horse on mars",
        "cute grey cat with blue eyes, wearing a bowtie, acrylic painting",
        "a cute magical flying dog, fantasy art drawn by disney concept artists, highly detailed, digital painting",
        "an illustration of a house with large barn with many cute flower pots and beautiful blue sky scenery",
        "one apple sitting on a table, still life, reflective, full color photograph, centered, close-up product",
        "background texture of stones, masterpiece, artistic, stunning photo, award winner photo",
        "new international organic style house, tropical surroundings, architecture, 8k, hdr",
        "beautiful Renaissance Revival Estate, Hobbit-House, detailed painting, warm colors, 8k, trending on Artstation",
        "blue owl, big green eyes, portrait, intricate metal design, unreal engine, octane render, realistic",
        "delicate elvish moonstone necklace on a velvet background, symmetrical intricate motifs, leaves, flowers, 8k",
    ]

    negative_prompt = "bad composition, ugly, abnormal, malformed"

    return prompts, negative_prompt

prompts, negative_prompt = example_prompts()

def inference(pipe, pipe_type:str):
    NUM_IMAGES_PER_PROMPT = 1
    steps = 30
    seed = 100
    latency_list = []

    for i, prompt in enumerate(prompts):
        inference_start = time.perf_counter()
        g = torch.Generator().manual_seed(seed)
        PIPELINE_GENERATION_CONFIG = {"prompt":prompt,
                                        "negative_prompt":negative_prompt,
                                        "num_inference_steps": steps,
                                        "num_images_per_prompt": NUM_IMAGES_PER_PROMPT,
                                        "guidance_scale": 7.5,
                                        "output_type": "pil",
                                        "generator": g
                                    }
        image = pipe(**PIPELINE_GENERATION_CONFIG).images[0]
        inference_end = time.perf_counter()
        latency = inference_end - inference_start
        latency_list.append(latency)
        print(f"End2End took {latency:.3f} seconds")
        image.save(f"./results/{pipe_type}/{pipe_type}_{i}.png")


inference(stable_diffusion, "optimum")

model_id = "runwayml/stable-diffusion-v1-5"
og_pipe = StableDiffusionPipeline.from_pretrained(model_id, torch_dtype=torch.float32)
og_pipe = og_pipe.to("cpu")

inference(og_pipe, "og")
```

|   | Prompt | original pipeline | neuron pipeline |
|---|--------|-------------------|-----------------|
| Row 1 | "a cute magical flying dog, fantasy art drawn by disney concept artists, highly detailed, digital painting" | ![og_2](https://github.com/huggingface/optimum-neuron/assets/97098931/bd5c7970-74a0-48de-bafa-6acf44105d0e) | ![optimum_2](https://github.com/huggingface/optimum-neuron/assets/97098931/97d63c21-8e5f-4f95-acf9-5045e6beb9f7) |
| Row 2 | "one apple sitting on a table, still life, reflective, full color photograph, centered, close-up product" | ![og_4](https://github.com/huggingface/optimum-neuron/assets/97098931/d46260e1-268b-428b-b3be-9857f6563c94) | ![optimum_4](https://github.com/huggingface/optimum-neuron/assets/97098931/f468a861-9fe9-4b45-bc7c-10c5d5542759) |
| Row 3 | "blue owl, big green eyes, portrait, intricate metal design, unreal engine, octane render, realistic" | ![og_8](https://github.com/huggingface/optimum-neuron/assets/97098931/69796b96-6fc1-4de0-babf-65d7699537fb) | ![optimum_8](https://github.com/huggingface/optimum-neuron/assets/97098931/300a57e4-4816-4f4f-9fb2-b5679cd7345f) |

### Expected behavior

The reason I wrote this post is because I want to use NeuronStableDiffusionPipeline to get the same or very similar results to the images generated by original StableDiffusionPipeline.

Is it common to have a different image than the original StableDiffuisionPipeline when using NeuronStableDiffuisonPipeline?

Or is this a problem that can be solved by adding certain options when compiling the NeuronStableDiffusionPipeline?

Please let me know how to generate the same image when using NeuronStableDiffusionPipeline.

**Comment by JingyaHuang** [comment]
Hi @newgrit1004,

Thanks for reporting it back to us!

I can reproduce the image quality issue, and after running some other experiments, it seems to be an issue that occurs only for the `runwayml/stable-diffusion-v1-5` checkpoint. This is the log I got with:

* cli

```
optimum-cli export neuron --model runwayml/stable-diffusion-v1-5 --task stable-diffusion --batch_size 1 --height 512 --width 512 --num_images_per_prompt 1 --auto_cast none sd_neuron_no_cast_15/
```

* log  

```
2024-05-23 16:57:19.000783:  20636  INFO ||NEURON_CACHE||: Compile cache path: /var/tmp/neuron-compile-cache
2024-05-23 16:57:19.000788:  20636  INFO ||NEURON_CACHE||: Compile cache path: /var/tmp/neuron-compile-cache
Model cached in: /var/tmp/neuron-compile-cache/neuronxcc-2.13.66.0+6dfecc895/MODULE_bdd74709b718bd5c0341.
Validating text_encoder model...
        - Validating Neuron Model output "last_hidden_state":
                -[✓] (1, 77, 768) matches (1, 77, 768)
                -[✓] all values close (atol: 0.001)
        - Validating Neuron Model output "pooler_output":
                -[✓] (1, 768) matches (1, 768)
                -[✓] all values close (atol: 0.001)
Validating unet model...
        -[✓] Neuron model output names match reference model (sample)
        - Validating Neuron Model output "sample":
                -[✓] (1, 4, 64, 64) matches (1, 4, 64, 64)
                -[x] values not close enough, max diff: 0.20612472295761108 (atol: 0.001)
The maximum absolute difference between the output of the reference model and the Neuron exported model is not within the set tolerance 0.001:
- sample: max diff = 0.20612472295761108
Validating vae_decoder model...
        -[✓] Neuron model output names match reference model (sample)
        - Validating Neuron Model output "sample":
                -[✓] (1, 3, 512, 512) matches (1, 3, 512, 512)
                -[✓] all values close (atol: 0.001)
The Neuronx export succeeded and the exported model was saved at: sd_neuron_no_cast_15
```

As you might observe, the result of unet on Neuron diverges quite significantly from the results on CPU even without casting the model's ops to low bits (all matmul ops to bf16 by default).

For example, this is what I saw while exporting the `stabilityai/stable-diffusion-2-1-base` checkpoint:
```
[Compilation Time] 257.91 seconds.
[Total compilation Time] 770.05 seconds.
2024-05-23 16:10:46.000967:  12600  INFO ||NEURON_CACHE||: Compile cache path: /var/tmp/neuron-compile-cache
2024-05-23 16:10:46.000971:  12600  INFO ||NEURON_CACHE||: Compile cache path: /var/tmp/neuron-compile-cache
Model cached in: /var/tmp/neuron-compile-cache/neuronxcc-2.13.66.0+6dfecc895/MODULE_4bdf0d3f9c9ce4d6bc3b.
Validating text_encoder model...
        - Validating Neuron Model output "last_hidden_state":
                -[✓] (1, 77, 1024) matches (1, 77, 1024)
                -[x] values not close enough, max diff: 0.0012061595916748047 (atol: 0.001)
        - Validating Neuron Model output "pooler_output":
                -[✓] (1, 1024) matches (1, 1024)
                -[✓] all values close (atol: 0.001)
The maximum absolute difference between the output of the reference model and the Neuron exported model is not within the set tolerance 0.001:
- last_hidden_state: max diff = 0.0012061595916748047
Validating unet model...
        -[✓] Neuron model output names match reference model (sample)
        - Validating Neuron Model output "sample":
                -[✓] (1, 4, 64, 64) matches (1, 4, 64, 64)
                -[x] values not close enough, max diff: 0.0031936615705490112 (atol: 0.001)
The maximum absolute difference between the output of the reference model and the Neuron exported model is not within the set tolerance 0.001:
- sample: max diff = 0.0031936615705490112
Validating vae_decoder model...
        -[✓] Neuron model output names match reference model (sample)
        - Validating Neuron Model output "sample":
                -[✓] (1, 3, 512, 512) matches (1, 3, 512, 512)
                -[✓] all values close (atol: 0.001)
``` 

And the generated images:

|   | Prompt | original pipeline | neuron pipeline(auto_cast: none) | neuron pipeline(auto_cast: matmul bf16) |
|---|--------|-------------------|-----------------|-----------------|
| Row 1 | "a cute magical flying dog, fantasy art drawn by disney concept artists, highly detailed, digital painting" | ![image](https://github.com/huggingface/optimum-neuron/assets/44135271/73417c36-acaf-47b9-ba24-82e203aa482c)|![image](https://github.com/huggingface/optimum-neuron/assets/44135271/2c013576-cc5e-4026-bf87-2e660d752559) |
| Row 2 | "one apple sitting on a table, still life, reflective, full color photograph, centered, close-up product" | ![image](https://github.com/huggingface/optimum-neuron/assets/44135271/73782e40-6dcc-41e7-b76b-a9a7c2c1e2e3) | ![image](https://github.com/huggingface/optimum-neuron/assets/44135271/17e78413-9a93-4f72-a7d7-654b721314bf)|
| Row 3 | "blue owl, big green eyes, portrait, intricate metal design, unreal engine, octane render, realistic" | ![image](https://github.com/huggingface/optimum-neuron/assets/44135271/63e8a095-a70f-4e1d-bc8a-556903211237) | ![image](https://github.com/huggingface/optimum-neuron/assets/44135271/0daaecaa-b70d-47f1-a6df-4b2b935f6fa2) |

(casting all matmul ops to bf16 gets almost the same quality as well)

Here are some of my guesses for the issue:
* In optimum-neuron, we previously override some ops to pass the compilation, maybe there is something not applicable for SD1.5.
* Maybe it's introduced by the neuron compiler

Let me check further and get back to you.

**Comment by newgrit1004** [comment]
Thank you for your quick response. I totally understand why compiled SD1.5 model is not working well.
Also, I can reproduce that the generated images are the same when using compiled SD 2.1 model.

I appreciate that you're trying to figure out the reason why compiled SD 1.5 model is not working well.
I hope this bug would be fixed as soon as possible.

**Comment by newgrit1004** [comment]
Thanks a million @JingyaHuang, I will wait until the PR is merged.
I also want to know why the optimized version was not working, so please let me know when you get the response from Annapurna team!

**Comment by JingyaHuang** [comment]
The SD15 quality with optimized attention score will be fixed with this PR: #646
